pkg/host/hugepages.go

```diff
@@ -0,0 +1,115 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package host
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+
+	"github.com/digitalocean/go-libvirt"
+	"github.com/go-logr/logr"
+	"github.com/ironcore-dev/libvirt-provider/pkg/api"
+	"github.com/ironcore-dev/libvirt-provider/pkg/controllers"
+	"libvirt.org/go/libvirtxml"
+)
+
+type HugepageTuner struct {
+	Lv           *libvirt.Libvirt
+	HugePageSize uint64
+}
+
+func InitHugepageTuner(lv *libvirt.Libvirt, HugePageSize uint64) (controllers.TuneFunc, error) {
+	nt := HugepageTuner{
+		Lv:           lv,
+		HugePageSize: HugePageSize,
+	}
+	return nt.Tune, nil
+}
+
+func (n *HugepageTuner) Tune(ctx context.Context, domain *libvirtxml.Domain, _ *api.Machine, _ *controllers.MachineReconciler) error {
+	log, err := logr.FromContext(ctx)
+	if err != nil {
+		return fmt.Errorf("can't get logger from context: %w", err)
+	}
+
+	if domain.Memory == nil {
+		return nil
+	}
+	requiredMemory := uint64(domain.Memory.Value)
+
+	if domain.Memory.Unit != Byte {
+		return fmt.Errorf("unsupported memory unit: %v, please set in bytes", domain.Memory.Unit)
+	}
+
+	log.V(1).Info("Hugepage memory tuning", "requiredMemory", requiredMemory, "domain", domain.Name)
+	if ok, err := n.isEnoughFreePagesAllocated(requiredMemory); !ok || err != nil {
+		if !ok {
+			return fmt.Errorf("not enough  memory in preallocated amount of hugepages ")
+		}
+		return fmt.Errorf("unable to get free pages info: %w", err)
+	}
+
+	domain.MemoryBacking = &libvirtxml.DomainMemoryBacking{
+		MemoryHugePages: &libvirtxml.DomainMemoryHugepages{},
+	}
+
+	return nil
+}
+
+func (n *HugepageTuner) isEnoughFreePagesAllocated(memory uint64) (bool, error) {
+	needMemoryKb, _ := AlignMemory(memory, 1024)
```

Comment:
optional: you use 1024 on multiple places. it will be better set as varialbe or  move into constants